File: codeflix/Catalog/src/Catalog.Domain/Validation/DomainValidation.cs
using Catalog.Domain.Exceptions;

namespace Catalog.Domain.Validation;

public class DomainValidation
{
    public static void NotNull<T>(T target, string fieldName)
    {
        if (target == null)
            throw new EntityValidationException($"{fieldName} should not be null");
    }

    public static void NotNullOrEmpty(string? target, string fieldName)
    {
        if (string.IsNullOrWhiteSpace(target))
            throw new EntityValidationException($"{fieldName} should not be empty or null");
    }

    public static void MinLength(string target, int minLength, string fieldName)
    {
        if (target.Length < minLength) 
            throw new EntityValidationException($"{fieldName} should be at least {minLength} characters long");
    }

    public static void MaxLength(string target, int maxLength, string fieldName)
    {
        if (target.Length > maxLength)
            throw new EntityValidationException($"{fieldName} should be less or equal {maxLength} characters long");
    }
}

File: codeflix/catalog_rust/src/catalog_domain/validation/domain_validation.rs
use crate::catalog_domain::EntityValidationError;

pub struct DomainValidation;

impl DomainValidation {
    pub fn not_empty_or_whitespace(
        value: &str,
        field_name: &str,
    ) -> Result<(), EntityValidationError> {
        if value.trim().is_empty() {
            return Err(EntityValidationError::new(Some(format!(
                "{} should not be empty or whitespace",
                field_name
            ))));
        }
        Ok(())
    }

    pub fn min_length(
        value: &str,
        field_name: &str,
        min_length: usize,
    ) -> Result<(), EntityValidationError> {
        if value.len() < min_length {
            return Err(EntityValidationError::new(Some(format!(
                "{} should be at least {} characters long",
                field_name, min_length
            ))));
        }
        Ok(())
    }
    pub fn max_length(
        value: &str,
        field_name: &str,
        max_length: usize,
    ) -> Result<(), EntityValidationError> {
        if value.len() > max_length {
            return Err(EntityValidationError::new(Some(format!(
                "{} should be less or equal {} characters long",
                field_name, max_length
            ))));
        }
        Ok(())
    }
}
